--- metacity-2.24.0/src/compositor/compositor-xrender.c.old	2009-07-02 20:14:02.951447316 +0200
+++ metacity-2.24.0/src/compositor/compositor-xrender.c	2009-07-02 20:15:12.530448817 +0200
@@ -1443,16 +1443,16 @@
   if (compositor->repaint_id > 0)
     return;
 
-#if 1
-  compositor->repaint_id = g_idle_add_full (G_PRIORITY_HIGH_IDLE,
-                                            compositor_idle_cb, compositor,
-                                            NULL);
-#else
-  /* Limit it to 50fps */
-  compositor->repaint_id = g_timeout_add_full (G_PRIORITY_HIGH, 20,
+//#if 1
+//  compositor->repaint_id = g_idle_add_full (G_PRIORITY_HIGH_IDLE,
+//                                            compositor_idle_cb, compositor,
+//                                            NULL);
+//#else
+  /* Limit it to 60ish fps */
+  compositor->repaint_id = g_timeout_add_full (G_PRIORITY_HIGH, 17,
                                                compositor_idle_cb, compositor,
                                                NULL);
-#endif
+//#endif
 }
 #endif
 
