--- src/gdevdljm.c  2007-04-26 15:05:28.000000000 +0200
+++ src/gdevdljm.c  2007-04-26 15:07:40.000000000 +0200
@@ -125,10 +125,16 @@
                fputs("\033&l1S", prn_stream);
        }
     }
-    /* Put out per-page initialization. */
-    if (features & PCL_CAN_SET_PAPER_SIZE){ 
-        fprintf(prn_stream, "\033&l%dA", paper_size); 
-    } 
+    /* Put out per-page initialization, unless we are duplexing.  In duplex 
+       mode setting the paper size causes a page eject.  It's not clear what 
+       the optimal thing to do in the case where the page size changes in the 
+       middle of a duplex job, we chose to ignore the page size change. */
+    if (pdev->PageCount > 0 && 
+  !((features & PCL_HAS_DUPLEX) && ((dupset && dup) || (!dupset) ) ) ) {
+        if (features & PCL_CAN_SET_PAPER_SIZE) { 
+            fprintf(prn_stream, "\033&l%dA", paper_size); 
+        }
+    }
     fputs("\033&l0o0l0E", prn_stream);
     fputs(page_init, prn_stream);
     fprintf(prn_stream, "\033&l%dX", num_copies);      /* # of copies */
